### PR TITLE
Fix/token holders

### PR DESCRIPTION
**Pull request by trinhdn97**
# Type of change

- [X] Count token holder as a address
- [X] Attempt to fix total transactions when boot
- [X] Fix internal transactions and logs in TxByHash API
- [X] Get KRC token info from RPC when insert/update

### CODE REVIEW

File: server/info.go [hunk 1]
```diff
@@ -174,8 +174,17 @@ func (s *infoServer) GetCurrentStats(ctx context.Context) uint64 {
...
 		zap.Uint64("TotalTransactions", stats.TotalTransactions), zap.Uint64("TotalAddresses", stats.TotalAddresses),
 		zap.Uint64("TotalContracts", stats.TotalContracts))
-	_ = s.cacheClient.SetTotalTxs(ctx, stats.TotalTransactions)
+	totalTxs, err := s.dbClient.TxsCount(ctx)
+	if err != nil {
+		s.logger.Warn("Cannot get total txs when boot", zap.Uint64("totalTxs", totalTxs), zap.Error(err))
+	}
+	_ = s.cacheClient.SetTotalTxs(ctx, totalTxs)
 	_ = s.cacheClient.UpdateTotalHolders(ctx, stats.TotalAddresses, stats.TotalContracts)
+	_ = s.dbClient.InsertAddress(ctx, &types.Address{
```
Comment (ngdlong91) [inline server/info.go:183]: Plz handle error instead ignore

File: server/info.go [hunk 2]
```diff
@@ -752,7 +761,31 @@ func (s *infoServer) storeEvents(ctx context.Context, logs []types.Log, blockTim
...
+	for _, holder := range holdersList {
+		_, err = s.dbClient.AddressByHash(ctx, holder.HolderAddress)
+		if err != nil {
+			code, _ := s.kaiClient.GetCode(ctx, holder.HolderAddress)
+			_ = s.dbClient.InsertAddress(ctx, &types.Address{
+				Address:       holder.HolderAddress,
+				BalanceString: new(big.Int).SetInt64(0).String(),
+				IsContract:    len(code) > 0,
+			})
+			numOfNewAddress++
```
Comment (ngdlong91) [inline server/info.go:775]: Same

File: server/info.go [hunk 3]
```diff
@@ -899,18 +935,38 @@ func (s *infoServer) getInternalTxs(ctx context.Context, log *types.Log) *types.
...
+				Time:        block.Time,
+				TxHash:      e.TxHash,
+				TxIndex:     e.TxIndex,
+				BlockHash:   block.Hash,
+				Index:       e.Index,
+				Removed:     e.Removed,
+			},
+		}, block.Time)
+		if err != nil {
+			s.logger.Warn("Cannot store events to db", zap.Error(err))
```
Comment (ngdlong91) [inline server/info.go:966]: Same

File: server/loader.go [hunk 4]
```diff
@@ -22,8 +22,16 @@ func (s *infoServer) LoadBootData(ctx context.Context) error {
...
+	totalTxs, err := s.dbClient.TxsCount(ctx)
+	if err != nil {
+		s.logger.Warn("Cannot get total txs when boot", zap.Uint64("totalTxs", totalTxs), zap.Error(err))
+	}
+	_ = s.cacheClient.SetTotalTxs(ctx, totalTxs)
 	_ = s.cacheClient.UpdateTotalHolders(ctx, stats.TotalAddresses, stats.TotalContracts)
+	_ = s.dbClient.InsertAddress(ctx, &types.Address{
+		Address:    "0x",
+		IsContract: false,
+	})
```
Comment (ngdlong91) [inline server/loader.go:35]: same

File: server/rest.go [hunk 5]
```diff
@@ -1271,29 +1284,33 @@ func (s *Server) InsertContract(c echo.Context) error {
...
-			}
-			if err == nil {
-				totalSupply, _ := s.kaiClient.GetKRCTotalSupply(ctx, &jsonABI, common.HexToAddress(addrInfo.Address))
-				if totalSupply != nil {
-					addrInfo.TotalSupply = totalSupply.String()
-				}
-			}
-		}
+	krcTokenInfoFromRPC, err := s.getKRCTokenInfoFromRPC(ctx, addrInfo.Address, addrInfo.KrcTypes)
+	if err != nil && strings.HasPrefix(addrInfo.KrcTypes, "KRC") {
```
Comment (ngdlong91) [inline server/rest.go:1288]: Condition looks wrong? maybe !strings.HasPrefix  ? And use this check here 
if krcTokenInfoFromRPC != nil

File: server/rest.go [hunk 6]
```diff
@@ -1313,55 +1330,44 @@ func (s *Server) UpdateContract(c echo.Context) error {
...
+		s.logger.Warn("Updating contract is not KRC type", zap.Any("smcInfo", addrInfo), zap.Error(err))
+		return api.Invalid.Build(c)
+	}
+	if krcTokenInfoFromRPC != nil {
```
Comment (ngdlong91) [inline server/rest.go:1347]: same

File: server/rest.go [hunk 7]
```diff
@@ -1313,55 +1330,44 @@ func (s *Server) UpdateContract(c echo.Context) error {
...
+		// cache new token info
+		krcTokenInfoFromRPC.Logo = addrInfo.Logo
+		_ = s.cacheClient.UpdateKRCTokenInfo(ctx, krcTokenInfoFromRPC)
```
Comment (ngdlong91) [inline server/rest.go:1350]: same

File: server/rest.go [hunk 8]
```diff
@@ -1313,55 +1330,44 @@ func (s *Server) UpdateContract(c echo.Context) error {
...
+		addrInfo.TokenSymbol = krcTokenInfoFromRPC.TokenSymbol
+		addrInfo.TotalSupply = krcTokenInfoFromRPC.TotalSupply
+		addrInfo.Decimals = krcTokenInfoFromRPC.Decimals
+		addrInfo.NumTokensTotal = krcTokenInfoFromRPC.NumTokensTotal
 	}
 	currTokenInfo, _ := s.dbClient.AddressByHash(ctx, addrInfo.Address)
 	if err := s.dbClient.UpdateContract(ctx, &contract, &addrInfo); err != nil {
 		lgr.Error("cannot bind insert", zap.Error(err))
 		return api.InternalServer.Build(c)
 	}
```
Comment (ngdlong91) [inline server/rest.go:1362]: nil currToken is ok ?

**Review by ngdlong91 — COMMENTED** [review]
Also, better with using go-kaiclient instead since we going to remove internal client soon.

File: server/rest.go (hunk 5 above)
Comment (trinhdn97) [reply server/rest.go:1288]: We only get info of KRC tokens (`strings.HasPrefix(addrInfo.KrcTypes, "KRC")`) here. If this contract is KRC and we got error when try to retrieve its info from network then better return `Invalid` before making any changes to db/cache.

File: server/rest.go (hunk 8 above)
Comment (trinhdn97) [reply server/rest.go:1362]: It's ok, I will checked it later at `if (currTokenInfo != nil && currTokenInfo.KrcTypes == "" && currTokenInfo.TokenName == "" && currTokenInfo.TokenSymbol == "") || currTokenInfo == nil {`

File: server/info.go (hunk 1 above)
Comment (trinhdn97) [reply server/info.go]: Done!

**Pull request merged**